File: backend/analyzer.py
import re
from pathlib import Path
from typing import Dict, List, Set, Any

MAX_PASSWORD_LENGTH = 256
COMMON_PASSWORDS_FILE = Path(__file__).parent.parent / "data" / "common_passwords.txt"

def load_common_passwords() -> Set[str]:
    """Loads the set of common passwords from file, returning a lowercased set."""
    common_set: Set[str] = set()
    if COMMON_PASSWORDS_FILE.exists():
        with open(COMMON_PASSWORDS_FILE, "r", encoding="utf-8", errors="ignore") as f:
            for line in f:
                cleaned = line.strip().lower()
                if cleaned:
                    common_set.add(cleaned)
    else:
        # Minimal embedded fallback set
        common_set = {
            "123456", "password", "123456789", "12345678", "12345", "1234567",
            "qwerty", "admin", "welcome", "iloveyou", "sunshine", "letmein"
        }
    return common_set

COMMON_PASSWORDS = load_common_passwords()


def has_sequential_or_repeated(password: str) -> bool:
    """
    Checks for 3+ consecutive repeated characters (e.g. 'aaa')
    or 3+ sequential characters (e.g. 'abc', 'cba', '123', '321').
    """
    if len(password) < 3:
        return False

    lowered = password.lower()
    for i in range(len(lowered) - 2):
        c1, c2, c3 = ord(lowered[i]), ord(lowered[i+1]), ord(lowered[i+2])

        # 3 identical consecutive characters
        if c1 == c2 == c3:
            return True

        # Ascending sequence (abc, 123)
        if c2 == c1 + 1 and c3 == c2 + 1:
            return True

        # Descending sequence (cba, 321)
        if c2 == c1 - 1 and c3 == c2 - 1:
            return True

    return False


def analyze_password(password: str) -> Dict[str, Any]:
    """
    Analyzes password strength using regex heuristics, character class checks,
    sequential character detection, and dictionary matching.
    """
    if not password:
        return {
            "score": 0,
            "label": "Very Weak",
            "checks": {
                "length_ok": False,
                "has_lower": False,
                "has_upper": False,
                "has_digit": False,
                "has_symbol": False,
                "is_common_password": False,
                "has_sequential_chars": False,
            },
            "suggestions": ["Enter a password to analyze its strength."]
        }

    # Truncate input if it exceeds max allowed length to prevent ReDoS / hashing DoS
    eval_password = password[:MAX_PASSWORD_LENGTH]

    checks = {
        "length_ok": len(eval_password) >= 12,
        "has_lower": bool(re.search(r"[a-z]", eval_password)),
        "has_upper": bool(re.search(r"[A-Z]", eval_password)),
        "has_digit": bool(re.search(r"\d", eval_password)),
        "has_symbol": bool(re.search(r"[^A-Za-z0-9]", eval_password)),
        "is_common_password": eval_password.lower() in COMMON_PASSWORDS,
        "has_sequential_chars": has_sequential_or_repeated(eval_password),
    }

    # Score calculation
    # Length contributes up to 60 points (3 points per character up to 20 chars)
    score = min(len(eval_password), 20) * 3

    # Character class completeness adds 10 points per class (up to 40)
    class_count = sum([checks["has_lower"], checks["has_upper"], checks["has_digit"], checks["has_symbol"]])
    score += class_count * 10

    # Penalties
    if checks["is_common_password"]:
        score = min(score, 10)  # Hard cap for known breach/dictionary passwords

    if checks["has_sequential_chars"]:
        score -= 15

    # Clamp score between 0 and 100
    score = max(0, min(100, score))

    # Rating label determination
    if score <= 20:
        label = "Very Weak"
    elif score <= 40:
        label = "Weak"
    elif score <= 60:
        label = "Fair"
    elif score <= 80:
        label = "Strong"
    else:
        label = "Very Strong"

    # Build actionable suggestions
    suggestions: List[str] = []
    if checks["is_common_password"]:
        suggestions.append("This is a commonly used password. Choose a unique passphrase.")
    if not checks["length_ok"]:
        suggestions.append("Increase password length to at least 12 characters (16+ recommended).")
    if not checks["has_lower"]:
        suggestions.append("Add lowercase letters (a-z).")
    if not checks["has_upper"]:
        suggestions.append("Add uppercase letters (A-Z).")
    if not checks["has_digit"]:
        suggestions.append("Include numeric digits (0-9).")
    if not checks["has_symbol"]:
        suggestions.append("Include special symbols (e.g. !@#$%^&*).")
    if checks["has_sequential_chars"]:
        suggestions.append("Avoid repetitive or sequential character patterns (e.g. 'abc', '123', 'aaa').")

    if not suggestions:
        suggestions.append("Great password! It satisfies all key security criteria.")

    return {
        "score": score,
        "label": label,
        "checks": checks,
        "suggestions": suggestions
    }

File: frontend/src/utils/analyzer.js
import { checkBloomFilter } from './bloomFilter';

const COMMON_WEAK_PASSWORDS = new Set([
  '123456', 'password', '123456789', '12345678', '12345', 'qwerty',
  '111111', '1234567', 'dragon', 'welcome', 'admin', 'master',
  'iloveyou', 'sunshine', 'princess', 'charlie', 'monkey', 'letmein'
]);

function checkSequential(str) {
  if (!str || str.length < 3) return false;
  const lower = str.toLowerCase();
  
  // Repeated characters (e.g. "aaa", "111")
  for (let i = 0; i < lower.length - 2; i++) {
    if (lower[i] === lower[i + 1] && lower[i] === lower[i + 2]) {
      return true;
    }
  }

  // Sequential char codes (e.g. "abc", "123")
  for (let i = 0; i < lower.length - 2; i++) {
    const code1 = lower.charCodeAt(i);
    const code2 = lower.charCodeAt(i + 1);
    const code3 = lower.charCodeAt(i + 2);
    if (code2 === code1 + 1 && code3 === code1 + 2) {
      return true;
    }
  }

  // Keyboard row patterns
  const keyboardPatterns = ['qwertyuiop', 'asdfghjkl', 'zxcvbnm'];
  for (const pattern of keyboardPatterns) {
    for (let i = 0; i <= pattern.length - 3; i++) {
      const sub = pattern.substring(i, i + 3);
      if (lower.includes(sub)) return true;
    }
  }

  return false;
}

/**
 * Evaluates password strength 100% in browser memory with zero network calls.
 */
export function analyzePasswordLocally(passwordInput) {
  if (!passwordInput) {
    return {
      score: 0,
      label: 'Awaiting Input',
      entropyBits: 0,
      checks: {
        length_ok: false,
        has_lower: false,
        has_upper: false,
        has_digit: false,
        has_symbol: false,
        is_common_password: false,
        has_sequential_chars: false
      },
      suggestions: ['Type a password to analyze strength in real-time.']
    };
  }

  const password = passwordInput.slice(0, 256);

  const has_lower = /[a-z]/.test(password);
  const has_upper = /[A-Z]/.test(password);
  const has_digit = /[0-9]/.test(password);
  const has_symbol = /[^a-zA-Z0-9]/.test(password);
  const length_ok = password.length >= 12;

  const is_common_password = checkBloomFilter(password) || COMMON_WEAK_PASSWORDS.has(password.toLowerCase());
  const has_sequential_chars = checkSequential(password);

  let poolSize = (has_lower ? 26 : 0) + (has_upper ? 26 : 0) + (has_digit ? 10 : 0) + (has_symbol ? 32 : 0);
  if (poolSize === 0) poolSize = 26;

  const entropyBits = Math.round(password.length * Math.log2(poolSize) * 100) / 100;

  // Base score based on entropy
  let score = Math.min(100, Math.round(entropyBits * 1.1));

  // Modifiers
  if (password.length >= 16) score += 15;
  else if (password.length >= 12) score += 10;

  if (has_sequential_chars) score -= 15;
  if (is_common_password) score = Math.min(score, 10);
  if (password.length < 8) score = Math.min(score, 20);

  score = Math.max(0, Math.min(100, score));

  // Label assignment
  let label = 'Very Weak';
  if (score >= 85) label = 'Very Strong';
  else if (score >= 70) label = 'Strong';
  else if (score >= 50) label = 'Fair';
  else if (score >= 30) label = 'Weak';

  // Suggestions
  const suggestions = [];
  if (!length_ok) {
    suggestions.push('Increase length to at least 12 characters (16+ recommended for high entropy).');
  }
  if (!has_upper) {
    suggestions.push('Add uppercase letters (A-Z) to widen the character search space.');
  }
  if (!has_lower) {
    suggestions.push('Include lowercase letters (a-z).');
  }
  if (!has_digit) {
    suggestions.push('Mix in numeric digits (0-9).');
  }
  if (!has_symbol) {
    suggestions.push('Add special symbols (!@#$%^&*) to resist dictionary attacks.');
  }
  if (is_common_password) {
    suggestions.push('CRITICAL: This password appears in common breach lists. Change it immediately.');
  }
  if (has_sequential_chars) {
    suggestions.push('Avoid sequential runs (e.g. 123, abc) or repeated character series.');
  }

  return {
    score,
    label,
    entropyBits,
    checks: {
      length_ok,
      has_lower,
      has_upper,
      has_digit,
      has_symbol,
      is_common_password,
      has_sequential_chars
    },
    suggestions
  };
}
